crates/fluvio-storage/src/segment.rs

```diff
@@ -133,6 +133,8 @@ where
                 );
                 match max_offset_opt {
                     Some(max_offset) => {
+                        // max_offset comes from HW, which could be greater than current segment end.
+                        let max_offset = std::cmp::min(max_offset, self.get_end_offset());
```

Comment:
change variable into something different since it will be easily confused with origin max_offset.  So maybe `effective_max_offset` or something like that